starters/next-drupal-starter/pages/api/preview.js

```diff
@@ -1,34 +1,54 @@
+import { globalDrupalStateAuthStores } from "../../lib/drupalStateContext";
+
 const preview = async (req, res) => {
   // Check the secret and next parameters
   // This secret should only be known to this API route and the CMS
   if (req.query.secret !== process.env.PREVIEW_SECRET || !req.query.slug) {
-    return res.status(401).json({ message: 'Invalid token' })
+    return res.redirect("/500", 401);
   }
 
-  // TODO - Fetch the headless CMS to check if the provided `slug` exists
-  // getPostBySlug would implement the required fetching logic to the headless CMS
-  // const post = await getPostBySlug(req.query.slug)
+  // returns the store that matches the locale found in the requested url
+  // or the only store if using a monolingual backend
+  const [store] = globalDrupalStateAuthStores.filter(({ defaultLocale }) => {
+    const regex = new RegExp(`/${defaultLocale}/`);
+    return defaultLocale ? regex.test(req.url) : true;
+  });
+
+  // get the object name from the slug
+  const regex = new RegExp(
+    `^(?:/${store.defaultLocale}/|/)(?<objectName>.*)/.*$`
```

Comment:
This works for our out of the box setup, but I worry that it will be too restrictive as people build actual sites. For example, this wouldn't support a path like `/about` or any case where the first path argument doesn't have a 1:1 relation with the objectName. 

To fix this, my best guess is that we'd need to update the module to pass the object name as an additional query parameter. Since the open redirect can be considered a vulnerability, I'm going to merge the current state of this, and create tickets for future refactoring.

Reply:
Created:
https://getpantheon.atlassian.net/browse/DB-2875
https://getpantheon.atlassian.net/browse/DB-2876